### DISCUSSION TITLE
Extend the link-property-form in CKEditor fields to allow link classes

### DISCUSSION THREAD

**Discussion by spackmat** [sulu/sulu#6788] [Ideas]
In https://github.com/sulu/sulu-demo/pull/77 I accidentally opened a subject that I want to present here for discussion.

I have some cases where content editors usually add classes (or types if you want) to inline links inside of their content. Think of things like marking a link with a spoiler- or trigger-warning so that the frontend can catch up that class (or other attribute) of the `a` element and show whatever the frontend people decide to show to the readers for that. Or the most simple case: Highlight a link generically using a class and the css/js decides how that is implemented.

Basically @alexander-schranz answered on that, that in Sulu the default link class feature of the CKEditor is disabled on purpose for the reason that in Sulu content managers should not have impact on presentation / design of a website. From my point of view, allowing content editors to select one or more of some predefined classes for a link is a semantic decision and not a presentational. I.e. "what I want my link to represent" vs. "what I want my link to look like/behave".

So what I would love is an extension to Sulu's own link-property-form for the CKEditor to allow content editors to select one or more classes of a configurable list of link classes just like the existing selection of the link target. Or more generic a way to extend that helpful link-property-form with custom form elements. (Or that Sulu just doesn't replace CKEditor's own link class feature with its opinionated and thus on purpose not extensible link property UI).

What do others think? Maybe we are the only content editors that use link classes to differentiate between several types of links

**Comment by spackmat** [comment]
Hey, I stumbled upon that issue again, because on that site, it turned out the editors solved their need by simply not using my provided blocks, but use an html block instead with the needed code.

This is what happens, when one tries to enforce things like links that must not have a class to prevent template designers from misusing those classes to allow editors to make presentational only content decisions. For now, as a result, I as a template designer cannot allow editors to give links inside the RTE a semantical class and they start to write whole HTML structures like akkordeons manually, copied from the pattern library.

**Reply by alexander-schranz** [reply]
If you allow HTML/Script Injection I think we can not do much here, as all kind of things via such general embed block can be missued if a content manager does. Only think which may work would be stricter about what kind of code you allow and what you don't allow, else that is always an option they will hack things into your System.

But I kind of not understand that its simpler for them to create a new HTML block instead of a new CTA block. How it can be faster for them to write the code instead of select a page? Or is that block too complex or they just don't know about it?

I don't understand the `akkordeons` part as you ar etalking here about give links a styling not about creating special `akkordeons`. Which sure something which normally should be of text editing.

**Reply by spackmat** [reply]
They just need a class attribute on their links, because the pattern library uses link classes to handle things like inline buttons, spoiler warning marks, whatever. No script injection or anything needed nor wanted at all. The CSS/JS of the site does the job, when it finds links with such classes.

But because they cannot give links inside the RTE a class, they don't use my provided blocks and the RTE at all, but instead insert an HTML block and copy the needed structure from the pattern library's example code and insert their content into it. For example whole accordeons, for which I provide an own block with RTEs for each element. But they have no other choice, when they want to use the link class based features, because the RTE does not allow setting classes on links.

So all I want is a class input on links inside the RTE, just as the standard implementation of CKEditor has. It would even be better to have an abstract extension mechanism for the link modal to hook in, but a simple text input would just work fine to satisfy such needs.

I don't see any problem with injections on anything, but a semantic information, what type of link that is.

**Reply by alexander-schranz** [reply]
You can still customize the CKEditor to your needs and add custom toolbar actions to your needs. https://github.com/sulu/sulu-demo/pull/77

> I don't see any problem with injections on anything, but a semantic information, what type of link that is.

I'm not aware of any semantic attributes for your usecase, or did I miss something we are here talking about css classes `<a href="" class="..">` for styling not for any other semantic attributes which are interpreted by screen reader or other tools differently?

**Reply by spackmat** [reply]
From my point of view, using classes to give elements a certain behaviour, is a semantical use. Like it is [said e.g. at MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/class):

> Though the specification doesn't put requirements on the name of classes, web developers are encouraged to use names that describe the semantic purpose of the element, rather than the presentation of the element.

Classes can be purely presentational, like most classes of bootstrap fir example, but others are not.

But that rather academic question isn't that relevant here, as I mentioned usecases, that are for now not possible in Sulu's RTE, because it replaced the default link functionality of CKEditir by an own one, wich lacks the possibility to define a class on a link.

I could build my own RTE-link functionality with all the logic for the different link types (internal, external, media). And maintain that. But, seriously, it would be less work for me to reimplement the whole site on another CMS, which does not stand in my way regarding the modeling of my content. I don't understand, why it is so important for you to not allow classes on links, enforcing valid usecases to find complex workarounds.

**Reply by alexander-schranz** [reply]
@spackmat its just how we think as provider of this free open source CMS how content should be structured inside templates. The main concept of Sulu are Blocks and not have a bloated Text Editor like other Systems. Also we are heavily avoding that content manager requires any technical knowledge which includes CSS classes to keep the UI easy, also as that things can not be easily migrated to a redesign and content lives longer as any design or styling.

Our text editor is still [extendable](https://github.com/sulu/sulu-demo/pull/77) and you can add new custom toolbar actions if your project really requires it. Most things which get in core are previously already tested in our own projects or lived before as a third party bundle. Like akawaka which created a text editor bundle which allow removing buttons for specific usecases, and have a even more lightweight editors.

But back to your problem, I always say content should be structured for the way it is meaningful also for blind people, which make it automatically also accessible for SEO Crawlers and even todays AI Agents easier. So lets have a look at this problem, for SEO tools there is no different between: `<a>Test</a> and <a class="some">Other</a>` for them. So how do we make things more meaningful when we create such texts over another text? We make them bold, and a strong tag make it more important for screen readers over a default text so `<a>Test<a> vs <a><strong>Test</strong></a>`. How to we make a text diff from the other part we put it into an own paragraph. So from semantical meaning for even screen readers something like `<p><a><strong>Test<strong></a></p>` is maybe the more technical correct semantic meaning instead of any just styling based option, and you can style it as a button (`p > a:only-child:has(strong:only-child)`). You could create a Toolbar action (Make Button) which is only available on a link (a, sulu-link) which basically add that tags to the selected link. There is no need for this todo things on the link toolbar action itself as it can live as its own ckeditor toolbar button. Still you can overwrite a JS file with webpack itself.

**Reply by spackmat** [reply]
> content lives longer as any design or styling.

Yes, I'm totally with you there. But giving content a class fulfills that in the first place. When our frontend team redesigns our design and styling, we can perfectly rely on that (from my point of view) semantic classes: No content changes needed (as long as the frontend team does not break BC on their class names), portability is fulfilled, everything is fine.

Take for example (there is more) the marking/hint for internal/external links: A link is internal for us, when it links to the same site (inside Sulu, that case is easy), but it is also considered internal, when it links to another site within our organization. Thus, external links are links to a URL, but not to some, that only a content manager knows. And it is up to our content managers to decide that, because our system cannot decide that programmatically (as it is constantly changing in our case).

> You could create a Toolbar action (Make Button) which is only available on a link (a, sulu-link) which basically add that tags to the selected link. There is no need for this todo things on the link toolbar action itself as it can live as its own ckeditor toolbar button.

I see several suboptimal ways to implement that marking in a semantical way (=screenreder/SEO perspective), none of them is really consensual accepted or standardized at the moment and the somehow "best way" changed in the past more than once. And now comes the twist: When I decide to use one of them and implement it as a button to wrap a somewhat semantical HTML structure around the link element, the RTE pins that code structure for the future and when the frontend team decides to change that structure in use, we need a migration. That is exactly the thing you wanted to avoid by forbidding the use of classes on links. Other way around: Marking a link with a class will survive changes of the frontend code and keeps the handling of the presentation and(!) behavior within the frontend domain.

I'm also with you in that the problem is more complex and the optimal way would be a link modal with a dropdown for a given list of types (defined by the template designer) and then the template (=renderer for Sulu link tags) would decide on how to render the different versions. Just like with the dropdown for the target attribute. **I would love to get that, don't get me wrong**. But until that better world exists, allowing my content managers to set their classes on links to fulfill their requirements would be enough and it would not do any harm to everything else. And it would prevent them to write HTML manually **to achieve compliance with the design system**, flushing every content abstraction down the toilet.

I'm repeating myself: For now, Sulu prevents me to implement our given design system correctly, which is build CMS agnostic on purpose. And if I had known about those limitations prior to my decision about the used CMS for that site, Sulu would simply have been ruled out. In our list of requirements for a CMS we have full control over the outputted HTML to implement the given design system. And that must-level requirement is violated here, unless I implement and(!) maintain an own RTE link handling against the build in one. Quite surprising for me by the way, as the used CKEditor normally allows link classes like most other CMS out there and your argumentation against it sounds very doctrinal theoretic for me.

P.S. In my opinion, avoiding a full RTE in the first place is always the better way for abstract content handling. In other systems we use markdown text-fields with markdown extra to add classes to links, if needed. Or we use a custom made semi WYSIWYG editor with strong content abstraction based on TipTap and some buttons for limited semantic markup. Or, for example, take a look at the [Hanna Code Dialog](https://processwire.com/modules/hanna-code-dialog/) module in ProcessWire, which is so helpful for content editors to work with content abstractions inside a RTE (also CKEditor in that case). Hanna Code is ProcessWire's flavor of shortcodes. Good solutions for content abstraction within text content fields exist out there. Forbidding classes on links is not one of them in my opinion.
